### ISSUE TITLE
Feature Request:  Provide option to reject a script if using unapproved packages

### ISSUE THREAD

**Issue by parmsam-pfizer** [pharmaverse/logrx#168]
### Feature Idea

Allow the user to select a "ignore", "warn", "error" with a "warn" default when running `aexecute()` on unapproved packages. The aexecute function could show the selected diagnostic message for unapproved package use to the user. This would be a good feature enhancement to the vignette on [logging unapproved packages](https://pharmaverse.github.io/logrx/articles/approved.html). I think this could be added in the part of the log_write function that checks for `log.rx.approved`: https://github.com/pharmaverse/logrx/blob/ef8428e7072a55f7b80b3287766143ba673b7a38/R/log.R#L233-L243 Curious about others thoughts on this.

**Comment by nicholas-masel** [comment]
I like the feature, but I'm wondering what the expected outcome is here?  

Function and package use is found after the script has executed.  Is the expectation is the script would error immediately when an unapproved function is used?

**Comment by parmsam-pfizer** [comment]
Yes, that's a great point you raise, @nicholas-masel. I think currently we can only support the feedback after script execution. That's what I was leaning towards. Not sure if we would have the ability to find the functions/packages used prior to executing the script.

**Comment by thomas-neitmann** [comment]
How about parsing the script prior to executing and then looking for all instances of `library()`, `require()` or `pkg::fun()` to discover any packages used? That way the script would not even get executed.

**Comment by thomas-neitmann** [comment]
So I found this to be an intriguing problem and went down the rabbit hole. Here's what I've come up with:

```r
library(testthat)

extract_used_pkgs <- function(script) {
  ast <- parse(script)
  
  .extract_used_pkgs <- function(expr) {
    if (!is.call(expr)) {
      return(character())
    }
    
    if (expr[[1L]] == quote(library) || expr[[1L]] == quote(require) || expr[[1L]] == quote(`::`)) {
      if (isFALSE(expr$character.only) || is.null(expr$character.only)) {
        return(as.character(expr[[2L]]))
      }
      stop("Do not set `character.only` to `TRUE` inside `library()` or `require()`", call. = FALSE)
    }
    
    unlist(lapply(expr, .extract_used_pkgs))
  }
  
  pkgs <- lapply(ast, .extract_used_pkgs)
  unique(unlist(pkgs))
}

test_that("packages attached with `library()` are detected", {
  script <- c(
    "library(dplyr)",
    "library(ggplot2)",
    "library(stringr)",
    "x <- 0",
    "y <- x + 1"
  )
  file <- tempfile()
  writeLines(script, file)
  
  expect_identical(extract_used_pkgs(file), c("dplyr", "ggplot2", "stringr"))
})

test_that("packages attached with `require()` are detected", {
  script <- c(
    "require(dplyr)",
    "require(ggplot2)",
    "require(stringr)",
    "x <- 0",
    "y <- x + 1"
  )
  file <- tempfile()
  writeLines(script, file)
  
  expect_identical(extract_used_pkgs(file), c("dplyr", "ggplot2", "stringr"))
})

test_that("packages loaded with `::` are detected", {
  script <- c(
    "adsl <- haven::read_sas('./data/adsl.sas7bdat')",
    "dplyr::select(adsl, USUBJID, AGE, SEX)"
  )
  file <- tempfile()
  writeLines(script, file)
  
  expect_identical(extract_used_pkgs(file), c("haven", "dplyr"))
})

test_that("packages loaded with `::` are detected", {
  script <- c(
    "library('magrittr')",
    "adsl <- haven::read_sas('./data/adsl.sas7bdat')",
    "adsl %>% dplyr::select(USUBJID, AGE, SEX) %>% tidyr::filter(SEX == 'M')",
    "pkg1::select(pkg2::filter(adsl, SEX == 'M'), USUBJID, AGE, SEX)"
  )
  file <- tempfile()
  writeLines(script, file)
  
  expect_identical(extract_used_pkgs(file), c("magrittr", "haven", "dplyr", "tidyr", "pkg1", "pkg2"))
})

test_that("error when `character.only = TRUE`", {
  script <- c(
    "pkg1 <- 'dplyr'",
    "pkg2 <- 'ggplot2'",
    "library(pkg1, character.only = TRUE)",
    "require(pkg2, character.only = TRUE)"
  )
  file <- tempfile()
  writeLines(script, file)
  
  expect_error(extract_used_pkgs(file))
})
```

**Comment by parmsam-pfizer** [comment]
Great suggestion! That's exactly what we discussed at our last logrx team meeting. Thanks for sharing your code! We'll add this to our roadmap.

**Comment by nicholas-masel** [comment]
@thomas-neitmann @parmsam-pfizer  We finally got around to this feature and it falls over base and default package usage.  More details here: https://github.com/pharmaverse/logrx/issues/222#issuecomment-2701880128

**Comment by nicholas-masel** [comment]
@kodesiba Should we close this?  Looks like we were ok without this functionality based on #222 ; https://github.com/pharmaverse/logrx/issues/222#issuecomment-2776322570

**Comment by nicholas-masel** [comment]
We explored two approaches for statically extracting package names from a script prior to execution:

**Option 1: Expression walking**
Walk the parsed expression tree recursively, matching calls to library(), require(), and ::. This is readable and works at a high level of abstraction, but requires a recursive walker to handle arbitrary nesting (e.g. library() calls inside if blocks, loops, or function bodies). Handling all possible call types adds complexity that outweighs the readability benefit.

**Option 2: Token-based parsing**
Use parse() + getParseData() to get a flat table of every token in the file. Filter by SYMBOL_FUNCTION_CALL (for library/require) and SYMBOL_PACKAGE (for ::). Since getParseData() flattens the entire AST regardless of nesting depth, all package references are captured without any recursion.

We went with Option 2 as it handles arbitrary nesting for free and is consistent with the existing get_used_functions() implementation which already uses getParseData().
